### ISSUE TITLE
Adding features to leaflet blurred location display (needs breaking up into separate issues)

### ISSUE THREAD

**Issue by mridulnagpal** [publiclab/leaflet-blurred-location-display#9]
@jywarren Shall we extend this as well. Can you create some issues, I would love to work on this!!

**Comment by jywarren** [comment]
Yes, let's do! Are there some we can copy over from the big planning issue
in leaflet-blurred-location?

I have one: add the ability to re-center the map via a geocoding search
field. It could be almost identical to the input system, but just recenter
the map of displayed locations. Make sense?

Also 2b on this has a lot: https://github.com/publiclab/plots2/issues/1416

**Comment by mridulnagpal** [comment]
@jywarren Can you give an example?

**Comment by jywarren** [comment]
Sure! Basically imagine using this library over at http://publiclab.org/people -- and imagine you want to enter `Kenya` and see mostly people around Kenya. There could be a search text input to type "kenya" and just like in the location input form for adding to your profile, it could autocomplete and show suggestions like "Nairobi, Kenya" for example. If you clicked that, it'd recenter the map and zoom in on Nairobi. 

Make sense? Thanks!!!

**Comment by mridulnagpal** [comment]
Isn't something like this already being done in leaflet blurred location?

**Comment by mridulnagpal** [comment]
@jywarren Can you give a code example?

**Comment by sagarpreet-chadha** [comment]
Hi @mridulnagpal !

I think we have various maps (including inline maps) on our website , so i think lets create a new library or may be use this library to create a library such that we can display all maps using this library **only** AND each map having its own unique feature like  : 
  * **/people map** with SEARCH + Recent Contributors  , 
  * **/profile** with FULL Screen And Near-by Users maybe ,
  *  a **new-layer map**  with all options to choose different layer ( #2486 ) ,
  *  etc..

_Basically making all different features in a library and giving the developer/end-user the options to use any combination of these features . Makes sense ?_

What do you think ?

**Comment by mridulnagpal** [comment]
@jywarren Let's continue with this, shall we?

**Comment by mridulnagpal** [comment]
@jywarren Any views?

**Comment by jywarren** [comment]
Sorry Mridul! Been traveling (again!). Looking over this now.

**Comment by jywarren** [comment]
So, the basic function is:

```
to display a collection of blurred locations (generated for example with Leaflet.BlurredLocation) at varying zoom levels.

When zooming to higher zoom levels, blurrier locations will not be shown, if they are not at a high enough precision to be useful. For example:

When zoomed in to level _____, locations with 0.000 or better precision will be shown.

But when zooming further to level ____, locations with 0.0000 may be hidden, since they may likely fall outside the viewport.

(reference https://github.com/publiclab/leaflet-blurred-location/issues/98)
```

I like what @sagarpreet-chadha is suggesting -- i think the ability to search the map is key.

**Comment by jywarren** [comment]
So does this make sense? That in this library, as you zoom in, some locations appear only within some zoom levels, that are close to the precision they were entered in? If they have lots of precision, they might appear in many zoom levels, but if they are very imprecise (highly blurred), they would be hidden once you zoom in too far. 

We could write a test to check for this -- so it would start the map up, go to a certain zoom level, and check what's visible, then zoom in further and check again, for 3 or 4 zoom levels. Then we could write the functions of this library to ensure that the markers are properly displayed for those given zoom levels. 

@sagarpreet-chadha and @stefannibrasil and @milaaraujo are working on API queries that may also be used to populate this map on the Public Lab website. @sagarpreet-chadha can help explain how it was done to 'refresh on drag' with a new query to the API on features like this:

https://publiclab.org/wiki/inline-maps

(although those don't seem to be refreshing?)

Another task might be to add a "loading" indicator -- a spinner -- while map refreshes are occurring, so you know it's still loading.

**Comment by sagarpreet-chadha** [comment]
Hi @jywarren ...dynamic fetching of new data on zooming or panning is still not solved in this issue https://github.com/publiclab/plots2/issues/1934 (and not done by me 😄) .

Lets confirm if @stefannibrasil and @milaaraujo are working on it ❓  
Else i will start work on it 👍 👍 👍 

Thanks !

**Comment by sagarpreet-chadha** [comment]
I have done that in leaflet-environmental-layers as of now 😄 .

**Comment by mridulnagpal** [comment]
@sagarpreet-chadha @jywarren How do we fetch all the location objects from the database?

**Comment by jywarren** [comment]
ah, sorry, that issue seems so long ago! Thanks!

**Comment by stefannibrasil** [comment]
Hi, everyone, for now we are not working on that, but we definitely can in the next weeks! What I am almost finishing is this #2790

**Comment by jywarren** [comment]
For starters i think you can use a set of example data that's in the same format, so look at functions here:

https://publiclab.org/api/docs/ and https://github.com/publiclab/plots2/blob/master/doc/API.md

and maybe use https://publiclab.org/api/srch/taglocations?srchString=41,-89

You can have this be an example file you save statically in the library, in fact, in the example.

**Comment by jywarren** [comment]
But we might also think about the API providing a standard GeoJSON output?

**Comment by mridulnagpal** [comment]
@jywarren Working on a prototype, will push a demo soon!!

**Comment by mridulnagpal** [comment]
Hey @jywarren I tried 
`        $.getJSON("https://publiclab.org/api/srch/taglocations?srchString=41,-89", function(data) {
        	console.log(data);
        });
`
And am getting the error: 

Failed to load https://publiclab.org/api/srch/taglocations?srchString=41,-89: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'null' is therefore not allowed access.

**Comment by icarito** [comment]
Sure, will do!

**Comment by icarito** [comment]
Hi!
Please test! I think widely enabling `Access-Control-Allow-Origin` could facilitate some spoofing so I've enabled it **only** for the `/api/` routes! I hope that is enough? Let me know if you want other routes with that header added!

**Comment by mridulnagpal** [comment]
@icarito Yeah thanks man it works!!

**Comment by mridulnagpal** [comment]
@jywarren Here is the prototype 
![blurreddisplay](https://user-images.githubusercontent.com/22350740/42677750-5c028fa2-869a-11e8-8a05-0da11e3d854a.gif)

**Comment by stefannibrasil** [comment]
hey, everyone, @milaaraujo and I would be happy to help with anything. Do you still need some method from the API that we can help with? Cheers!

**Comment by mridulnagpal** [comment]
@stefannibrasil The CORS error was resolved and now I can use the API, thanks :)

**Comment by jywarren** [comment]
Ooh, this is really nice @mridulnagpal -- do you have a PR open for this? Do you think you could configure your fork's `gh-pages` branch and push this code there so we can try it out at https://mridulnagpal.github.io/Leaflet.BlurredLocationDisplay/examples/ ?

Great work!

**Comment by mridulnagpal** [comment]
@jywarren Done, you can test it here!! 
https://mridulnagpal.github.io/leaflet-blurred-location-display/examples/

**Comment by jywarren** [comment]
Cool! I'm not seeing any markers, though. And there does seem to be an old error, maybe we're not using the latest version here? 

![image](https://user-images.githubusercontent.com/24359/43106627-8148f20a-8ea7-11e8-8cc5-ba1218a600b5.png)

But good start, exciting!

**Comment by mridulnagpal** [comment]
@jywarren I am directly installing from npm, should we update the version there?

**Comment by mridulnagpal** [comment]
@jywarren #10 will be done once the API is updated, what should we add till then?

**Comment by jywarren** [comment]
OK i'm publishing a new version on npm! But, where do I look for the markers? Should I be seeing them when I start it up?

**Comment by jywarren** [comment]
published!

**Comment by jywarren** [comment]
Can we put a list of sample locations in the `example/index.html` file to show how this works, in the demo? And is this code now in the master branch? Can you push it into the publiclab/gh-pages branch too so we can show the demo there? Thanks!

**Comment by jywarren** [comment]
And with demo locations in the example file, we can move forward on #10, too, right? Thanks, Mridul!!!

**Comment by mridulnagpal** [comment]
@jywarren Do you mean move to a location by default where there are already tags?

**Comment by mridulnagpal** [comment]
Got it, working on it!!

**Comment by mridulnagpal** [comment]
@jywarren What's next?

**Comment by mridulnagpal** [comment]
@jywarren Should we move to something new as well?

**Comment by jywarren** [comment]
Yes, as soon as you're getting #13 working, i think you should think about how each grid square could be shaded darker based on how many blurred locations are within it -- want to think a little bit about how that could work? Would a function go through the visible grid squares and count, then establish a scale from "none" to "many" and shade accordingly, each time `setZoom` is run?
